lib/scholar/metrics.ex

```diff
@@ -98,197 +87,221 @@ defmodule Scholar.Metrics do
+
+    true_positives
+    |> Nx.divide(true_positives + false_positives + 1.0e-16)
+  end
+
+  @doc ~S"""
+  Computes the recall of the given predictions with respect to
+  the given targets for binary classification problems.
+
+  ## Argument Shapes
+
+    * `y_true` - $\(d_0, d_1, ..., d_n\)$
+    * `y_pred` - $\(d_0, d_1, ..., d_n\)$
 
   ## Examples
 
-      iex> Scholar.Metrics.recall(Nx.tensor([0, 1, 1, 1]), Nx.tensor([1, 0, 1, 1]))
+      iex> Scholar.Metrics.binary_recall(Nx.tensor([0, 1, 1, 1]), Nx.tensor([1, 0, 1, 1]))
       #Nx.Tensor<
         f32
         0.6666666865348816
       >
 
   """
-  defn recall(y_true, y_pred, opts \\ []) do
+  defn binary_recall(y_true, y_pred) do
     assert_shape(y_true, Nx.shape(y_pred))
 
-    true_positives = true_positives(y_true, y_pred, opts)
-    false_negatives = false_negatives(y_true, y_pred, opts)
+    true_positives = binary_true_positives(y_true, y_pred)
+    false_negatives = binary_false_negatives(y_true, y_pred)
 
     Nx.divide(true_positives, false_negatives + true_positives + 1.0e-16)
   end
 
-  @doc """
-  Computes the number of true positive predictions with respect
-  to given targets.
+  @doc ~S"""
+  Computes the recall of the given predictions with respect to
+  the given targets for multi-class classification problems.
+
+  ## Argument Shapes
+
+    * `y_true` - $\(d_0, d_1, ..., d_n\)$
+    * `y_pred` - $\(d_0, d_1, ..., d_n\)$
 
   ## Options
 
-    * `:threshold` - threshold for truth value of predictions.
-      Defaults to `0.5`.
+    * `:num_classes` - Number of classes contained in the input tensors
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([0, 1, 1, 1, 1, 0, 2, 1, 0, 1], type: {:u, 32})
+      iex> y_pred = Nx.tensor([0, 2, 1, 1, 2, 2, 2, 0, 0, 1], type: {:u, 32})
+      iex> Scholar.Metrics.recall(y_true, y_pred, num_classes: 3)
+      #Nx.Tensor<
+        f32[3]
+        [0.6666666865348816, 0.5, 1.0]
+      >
+
+  """
+  defn recall(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, [:num_classes])
+    assert_shape_pattern(y_true, {_})
+    assert_shape(y_pred, Nx.shape(y_true))
+
+    cm = confusion_matrix(y_true, y_pred, opts)
+    true_positive = Nx.take_diagonal(cm)
+    false_negative = Nx.subtract(Nx.sum(cm, axes: [1]), true_positive)
+    Nx.divide(true_positive, true_positive + false_negative + 1.0e-16)
+  end
+
+  @doc """
+  Computes the number of true positive predictions with respect
+  to given targets for binary classification problems.
 
   ## Examples
 
+      iex> threshold = 0.5
       iex> y_true = Nx.tensor([1, 0, 1, 1, 0, 1, 0])
       iex> y_pred = Nx.tensor([0.8, 0.6, 0.4, 0.2, 0.8, 0.2, 0.2])
-      iex> Scholar.Metrics.true_positives(y_true, y_pred)
+      iex> y_pred = Nx.greater(y_pred, threshold)
+      iex> Scholar.Metrics.binary_true_positives(y_true, y_pred)
       #Nx.Tensor<
         u64
         1
       >
   """
-  defn true_positives(y_true, y_pred, opts \\ []) do
+  defn binary_true_positives(y_true, y_pred) do
```

Comment:
Let's make these `defnp` :)

Reply:
oops, didn't notice that they break the `running_sum/1` doctest. I changed the doctest `running_sum/1` from using  `binary_false_positives/2` to `mean_absolute_error/2`. (although it seems not quite make sense to have a sum of mean absolute error...)